File: android/controller/src/main/java/org/openbot/controller/PhoneSensorToDualDriveConverter.kt
package org.openbot.controller

class PhoneSensorToDualDriveConverter  {
    private val g = 9.81f

    fun convert(
        azimuth: Float?,
        pitch: Float?,
        roll: Float?
    ): DualDriveValues {
        var leftSpeed = 0f
        var rightSpeed = 0f
        var forwardSpeed = 0f

        if (inDeadZone(roll!!)) {
            return DualDriveValues(0f, 0f)
        }

        // get forward speed
        forwardSpeed = ForwardSpeed.value

        // adjust for turning
        leftSpeed = forwardSpeed + (pitch?.div(g / 2) ?: 0f) * forwardSpeed
        rightSpeed = forwardSpeed - (pitch?.div(g / 2) ?: 0f) * forwardSpeed

        return DualDriveValues(leftSpeed, rightSpeed)
    }

    inner class DualDriveValues(var left: Float, var right: Float) {
        private val MAX = 1.0f
        private val MIN = -1.0f

        init {
            left = clean(left)
            right = clean(right)
        }

        private fun clean(value: Float): Float {
            var ret = value

            if (value > MAX) {
                ret = MAX
            }
            if (value < MIN) {
                ret = MIN
            }
            return ret.round(3)
        }

        private fun Float.round(decimals: Int): Float {
            var multiplier = 1.0f
            repeat(decimals) { multiplier *= 10 }
            return kotlin.math.round(this * multiplier) / multiplier
        }

        fun reset() {
            left = 0f
            right = 0f
        }
    }

    fun inDeadZone(roll: Float): Boolean {
        return isWithin(roll, 0f, 1f)
    }

    private fun isWithin(value: Float?, desiredValue: Float?, tolerance: Float?): Boolean {
        return value!! in (desiredValue!! - tolerance!!)..(desiredValue + tolerance)
    }
}

File: controller/flutter/lib/utils/phoneSensorToDualDriveConverter.dart
import 'package:openbot_controller/utils/forwardSpeed.dart';

class PhoneSensorToDualDriveConverter {
  static double g = 9.81;

  DualDriveValues convert(double azimuth, double pitch, double roll) {
    double leftSpeed = 0;
    double rightSpeed = 0;
    double forwardSpeed = 0;

    forwardSpeed = ForwardSpeed.value;
    leftSpeed = forwardSpeed + (pitch / (g / 2)) * forwardSpeed;
    rightSpeed = forwardSpeed - (pitch / (g / 2)) * forwardSpeed;
    return DualDriveValues(leftSpeed, rightSpeed);
  }
}

class DualDriveValues {
  double MAX = 1.0;
  double MIN = -1.0;
  double left = 0.0;
  double right = 0.0;

  DualDriveValues(double left, double right) {
    this.left = clean(left);
    this.right = clean(right);
  }

  double clean(double value) {
    double ret = value;
    if (value > MAX) {
      ret = MAX;
    }
    if (value < MIN) {
      ret = MIN;
    }
    return ret;
  }
}
